src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.Accessor.cs

```diff
@@ -156,57 +150,71 @@ public Task<Stream> ReadStreamAsync(TKey key, Checksum checksum, CancellationTok
+                        // only safe if we are in a transaction and no-one else can race with
+                        // us.
+                        return connection.RunInTransaction(
+                            ValidateChecksumAndReadBlob,
+                            (self: this, connection, writeCacheDB, columnName, checksumOpt, rowId, cancellationToken));
+                    }
+                }
+                catch (Exception ex)
+                {
+                    StorageDatabaseLogger.LogException(ex);
+                }
+
+                return null;
+
+                static Stream ValidateChecksumAndReadBlob(
+                    (Accessor<TKey, TWriteQueueKey, TDatabaseId> self, SqlConnection connection, bool writeCacheDB,
+                     string columnName, Checksum checksumOpt, long rowId, CancellationToken cancellationToken) t)
                 {
-                    // Have to run the blob reading in a transaction.  This is necessary
-                    // for two reasons.  First, blob reading outside a transaction is not
-                    // safe to do with the sqlite API.  It may produce corrupt bits if 
-                    // another thread is writing to the blob.  Second, if a checksum was
-                    // passed in, we need to validate that the checksums match.  This is
-                    // only safe if we are in a transaction and no-one else can race with
-                    // us.
-                    return connection.RunInTransaction((tuple) =>
+                    // If we were passed a checksum, make sure it matches what we have
+                    // stored in the table already.  If they don't match, don't read
+                    // out the data value at all.
+                    if (t.checksumOpt != null)
                     {
-                        // If we were passed a checksum, make sure it matches what we have
-                        // stored in the table already.  If they don't match, don't read
-                        // out the data value at all.
-                        if (tuple.checksumOpt != null &&
-                            !ChecksumsMatch_MustRunInTransaction(tuple.connection, tuple.rowId, tuple.checksumOpt, cancellationToken))
+                        if (!t.self.ChecksumsMatch_MustRunInTransaction(
+                                t.connection, t.writeCacheDB, t.rowId,
+                                t.checksumOpt, t.cancellationToken))
                         {
                             return null;
                         }
+                    }
 
-                        return connection.ReadBlob_MustRunInTransaction(tuple.self.DataTableName, tuple.columnName, tuple.rowId);
-                    }, (self: this, connection, columnName, checksumOpt, rowId));
+                    return t.connection.ReadBlob_MustRunInTransaction(t.writeCacheDB, t.self.DataTableName, t.columnName, t.rowId);
                 }
```

Comment:
move complex lambdas that we don't want to capture to static local functions.  we had bugs where we were still capturing accidentally (including 'this' and other values) ins several of these lambdas.